cmd/snap/inhibit.go

```diff
@@ -34,143 +34,160 @@ import (
+	hintFlock.Close()
+
+	if notified {
+		if err := flow.FinishInhibitionNotification(ctx); err != nil {
 			return err
 		}
-		if notifiedDesktopIntegration {
-			return nil
-		}
-		return graphicalSessionFlow(snapName, hint)
 	}
-	// terminal and headless
-	return textFlow(snapName, hint)
+
+	return nil
+}
+
+type inhibitionFlow interface {
+	StartInhibitionNotification(ctx context.Context) error
+	FinishInhibitionNotification(ctx context.Context) error
 }
 
-func inhibitMessage(snapName string, hint runinhibit.Hint) string {
-	switch hint {
-	case runinhibit.HintInhibitedForRefresh:
-		return fmt.Sprintf(i18n.G("snap package %q is being refreshed, please wait"), snapName)
-	default:
-		return fmt.Sprintf(i18n.G("snap package cannot be used now: %s"), string(hint))
+var newInhibitionFlow = func(instanceName string) inhibitionFlow {
+	if isGraphicalSession() {
+		return &graphicalFlow{instanceName: instanceName}
 	}
+	return &textFlow{instanceName: instanceName}
 }
 
-var isGraphicalSession = func() bool {
-	// TODO: uncomment once there is a proper UX review
-	//return os.Getenv("DISPLAY") != "" || os.Getenv("WAYLAND_DISPLAY") != ""
-	return false
+type textFlow struct {
+	instanceName string
+
+	pb progress.Meter
 }
 
-var pendingRefreshNotification = func(refreshInfo *client.PendingSnapRefreshInfo) error {
-	userclient := client.NewForUids(os.Getuid())
-	if err := userclient.PendingRefreshNotification(context.TODO(), refreshInfo); err != nil {
+func (tf *textFlow) StartInhibitionNotification(ctx context.Context) error {
+	_, err := fmt.Fprintf(Stdout, i18n.G("snap package %q is being refreshed, please wait\n"), tf.instanceName)
+	if err != nil {
 		return err
 	}
+	tf.pb = progress.MakeProgressBar(Stdout)
+	tf.pb.Spin(i18n.G("please wait..."))
 	return nil
 }
 
-var finishRefreshNotification = func(refreshInfo *client.FinishedSnapRefreshInfo) error {
-	userclient := client.NewForUids(os.Getuid())
-	if err := userclient.FinishRefreshNotification(context.TODO(), refreshInfo); err != nil {
-		return err
+func (tf *textFlow) FinishInhibitionNotification(ctx context.Context) error {
+	if tf.pb != nil {
+		tf.pb.Finished()
 	}
 	return nil
 }
 
-var tryNotifyRefreshViaSnapDesktopIntegrationFlow = func(snapName string) (bool, error) {
+type graphicalFlow struct {
+	instanceName string
+
+	notifiedDesktopIntegration bool
+}
+
+func (gf *graphicalFlow) StartInhibitionNotification(ctx context.Context) error {
+	gf.notifiedDesktopIntegration = tryNotifyRefreshViaSnapDesktopIntegrationFlow(ctx, gf.instanceName)
+	if gf.notifiedDesktopIntegration {
+		return nil
+	}
+
+	// unable to use snapd-desktop-integration, let's fallback graphical session flow
+	refreshInfo := client.PendingSnapRefreshInfo{
+		InstanceName: gf.instanceName,
+		// remaining time = 0 results in "Snap .. is refreshing now" message from
+		// usersession agent.
+		TimeRemaining: 0,
+	}
+	return pendingRefreshNotification(ctx, &refreshInfo)
```

Comment:
great catch, yes, I it will trigger. I believe it was already doing that before. This PR is just aimed at restructuring the notification flow to be usable with `runinhibit.WaitWhileInhibited`, I will definitely keep this in mind during the ongoing RAA UX flow review.

Reply:
I looked a bit further, it is unlikely but possible that `pendingRefreshNotification` from snap-run triggers at the same time another was issued from autorefresh.go.

This is a great observation because this is most definitely the case for `finishRefreshNotification` which is triggered after the snap is linked from both snap-run and snapd so I looked a bit deeper to check how problematic is this since I am doing a UX review for RAA in parallel currently.

Luckily the UI is not affected by this because both snap-run and snapd call userd to delegate the notification which ends up using the same notification ID for the snap so if those notifications trigger at the same time they will seamlessly replace each other.

- https://wiki.gnome.org/Projects/GLib/GApplication/DBusAPI#org.gtk.Notification
> The second parameter (s) is the notification id; if a notification with the same id already exists, it is replaced.
- https://github.com/snapcore/snapd/blob/e016f71207a67b622a0570f43aca47e97b8263b9/usersession/agent/rest_api.go#L422
- https://github.com/snapcore/snapd/blob/e016f71207a67b622a0570f43aca47e97b8263b9/usersession/agent/rest_api.go#L469